src/osd/ECBackend.cc

```diff
@@ -1181,11 +1186,9 @@ void ECBackend::handle_sub_read_reply(
         have.insert(j->first.shard);
         dout(20) << __func__ << " have shard=" << j->first.shard << dendl;
       }
-      set<int> want_to_read;
       map<int, vector<pair<int, int>>> dummy_minimum;
-      get_want_to_read_shards(&want_to_read);
       int err;
-      if ((err = ec_impl->minimum_to_decode(want_to_read, have, &dummy_minimum)) < 0) {
+      if ((err = ec_impl->minimum_to_decode(rop.want_to_read[iter->first], have, &dummy_minimum)) < 0) {
 	dout(20) << __func__ << " minimum_to_decode failed" << dendl;
         if (rop.in_progress.empty()) {
 	  // If we don't have enough copies and we haven't sent reads for all shards
```

Comment:
This comment is misleading during recovery when multiple pg_shard_t may be a available for a given shard_id_t (see get_remaining_shards()).  So send_all_remaining_shards() doesn't send multiple reads for the same shard_id_t.  So it is possible to get here multiple times if there are more errors in recovery only.  It is confusing because of the term "all shards" here doesn't apply to pg_shard_it but rather shard_id_t.  Saying "the rest of the reads" doesn't mean sending all possible reads for shards (shard_id_t) we still need.

Under normal read scenarios the shard_id_t to pg_shard_t is 1 to 1.

Reply:
good point, I'll fix the comment